src/protocol/mqtt.rs

```diff
@@ -0,0 +1,205 @@
+                            "heavy",
+                            "test"
+                        ],
+                        "type": "string",
+                        "forms": [
+                            {
+                                "href": "mqtt://broker.com",
+                                "mqv:filter": "application/deviceid/sensor/gas",
+                                "op": "observeproperty"
+                            }
+                        ]
+                    },
+                    "valve": {
+                        "title": "Valve",
+                        "observable": true,
+                        "enum": [
+                            "unknown",
+                            "closed",
+                            "opened",
+                            "not_connected",
+                            "failure",
+                            "closing",
+                            "opening",
+                            "checking"
+                        ],
+                        "type": "string",
+                        "forms": [
+                            {
+                                "href": "mqtt://broker.com/",
+                                "mqv:filter": "application/deviceid/valve/0/state",
+                                "op": "observeproperty"
+                            }
+                        ]
+                    }
+                },
+                "actions": {
+                    "mute": {
+                        "title": "Mute",
+                        "forms": [
+                            {
+                                "href": "mqtt://broker.com/",
+                                "mqv:topic": "application/deviceid/sensor/mute"
+                            }
+                        ]
+                    },
+                    "unmute": {
+                        "title": "Unmute",
+                        "forms": [
+                            {
+                                "href": "mqtt://broker.com/",
+                                "mqv:topic": "application/deviceid/sensor/unmute"
+                            }
+                        ]
+                    },
+                    "changeValveStatus": {
+                        "title": "Change Valve Status",
+                        "input": {
+                            "type": "string",
+                            "enum": [
+                                "open",
+                                "close"
+                            ]
+                        },
+                        "forms": [
+                            {
+                                "href": "mqtt://broker.com/",
+                                "mqv:topic": "application/deviceid/valve/0/command"
+                            }
+                        ]
+                    }
+                }
+            }
+        "#;
+
+        let f: crate::thing::Thing<Nil> = serde_json::from_str(s).unwrap();
```

Comment:
Same reasoning as before: if you want to test "just one giant thing" it's fine, but then you need to make an `assert_eq!` for what you expect, without testing the `Thing` and its _extendabiliy_ behavior, which is tested elsewhere.

Reply:
> You need to implement `PartialEq`, eventual `Eq` and `Hash` for all the structures. You also need to use `skip_serializing_none` or `serde(skip_serializing_if = "Option::None")` in some situations.

Yes,

> As already written, I totally don't like the current testing approach, which is not really evaluating the results of the deserialization and vaguely tries to check that the extension behavior in `wot-td` is working. For the former, proper assertions are needed, and for the latter it is not something that needs to be done in this context.

I copied from the documentation examples the snippets to see if what is written matched, I managed to catch faults in the examples themselves, I agree that the dbg! should be replaced with proper assertions.

Reply:
Surely enough, bogus examples don't help... :confused: